File: Comparative/testing/kotlinImp/lessThan20/problem72_countingFractions.kt
// https://projecteuler.net/problem=72
// (<20)

import kotlin.math.*

class CountingFractions {

    fun euler_phi(n: Long): Long {
        var newN = n
        var result = newN
        var p = 2L
        while (p * p <= newN) {
            if (newN % p == 0L) {
                while (newN % p == 0L) {
                    newN /= p
                }
                result -= result / p
            }
            p++
        }
        if (newN > 1) {
            result -= result / newN
        }
        return result
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val countingFractions = CountingFractions()
            var count = 0L
            for (d in 2L..1000000L) {
                count += countingFractions.euler_phi(d)
            }
            println(count)
        }
    }
}

File: Comparative/testing/goImp/lessThan20/problem72_countingFractions.go
package main

import "fmt"

func eulerPhi(n int) int {
	result := n
	p := 2
	for p*p <= n {
		if n%p == 0 {
			for n%p == 0 {
				n /= p
			}
			result -= result / p
		}
		p++
	}
	if n > 1 {
		result -= result / n
	}
	return result
}

func main() {
	count := 0
	for d := 2; d <= 1000000; d++ {
		count += eulerPhi(d)
	}
	fmt.Println(count)
}
